plotly/tools.py

```diff
@@ -2509,6 +2561,205 @@ def create_dendrogram(X, orientation="bottom", labels=None,
         return {'layout': dendrogram.layout,
                 'data': dendrogram.data}
 
+    @staticmethod
+    def create_annotated_heatmap(z, x=None, y=None, annotation_text=None,
+                                 colorscale=None, fontcolor=None,
+                                 showscale=False, reversescale=False,
+                                 **kwargs):
+        """
+        BETA function that creates annotated heatmaps
+
+        This functions adds annotations to each cell of the heatmap.
+
+        :param (list[list]) z: z matrix to create heatmap.
+        :param (list) x: x axis labels.
+        :param (list) y: y axis labels.
+        :param (list[list]) text: Text strings for annotations. Should be the
+            same dimmensions as the z matrix. If no text is added, the the
+            values of the z matrix are annotated. Default = z matrix values.
+        :param (list) fontcolor: List of two colorstrings: [min_text_color,
+            max_text_color] where min_text_color is applied to annotations for
+            heatmap values < (max_value - min_value)/2. If fontcolor is not
+            defined the colors are defined logically as black or white
+            depending on the heatmap's colorscale.
+        :param (bool) showscale: Display colorscale. Default = False
+        :param kwargs: kwargs passed through plotly.graph_objs.Heatmap.
+            These kwargs describe other attributes about the annotated Heatmap
+            trace such as the colorscale. For more information on valid kwargs
+            call help(plotly.graph_objs.Heatmap)
+
+        Example 1: Simple annotated heatmap with default configuration
+        ```
+        import plotly.plotly as py
+        from plotly.tools import FigureFactory as FF
+
+        z = [[0.300000, 0.00000, 0.65, 0.300000], 
+             [1, 0.100005, 0.45, 0.4300],
+             [0.300000, 0.00000, 0.65, 0.300000],
+             [1, 0.100005, 0.45, 0.00000]]
+
+        hm = FF.create_annotated_heatmap(z)
+        py.iplot(hm)
+        ```
+        """
+        # TODO: protected until #282
+        from plotly.graph_objs import graph_objs
+
+        # Avoiding mutables in the call signature
+        colorscale = colorscale if colorscale is not None else 'RdBu'
+        fontcolor = fontcolor if fontcolor is not None else []
+        FigureFactory._validate_annotated_heatmap(z, annotation_text)
+        annotations = _AnnotatedHeatmap(z, x, y, annotation_text,
+                                        colorscale, fontcolor, reversescale,
+                                        **kwargs).make_annotations()
+
+        if x or y:
```

Comment:
Doesn't seem like there's validation on the x and y values at this point, is there? If x or y are `0` or '' or `[]` (etc), this conditional will fall to the `else` block. Should we actually be failing on these cases to prevent strange bugs in the future?